homeassistant/components/transmission/coordinator.py

```diff
@@ -0,0 +1,336 @@
+        self.tm_api = api
+        self._tm_data = TransmissionData(hass, entry, api)
+        super().__init__(
+            hass,
+            name="Transmission",
+            logger=_LOGGER,
+            update_interval=timedelta(
+                seconds=entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+            ),
+        )
+
+    @property
+    def api(self) -> TransmissionData:
+        """Return the TransmissionData object."""
+        return self._tm_data
+
+    async def _async_update_data(self) -> SessionStats:
+        """Update transmission data."""
+        return await self.hass.async_add_executor_job(self.api.update)
+
+    async def async_setup(self) -> None:
+        """Set up the Transmission client."""
+        await self.hass.async_add_executor_job(self.api.init_torrent_list)
+        self.add_options()
+        self.config_entry.add_update_listener(self.async_options_updated)
+
+    def register_services(self) -> None:
+        """Register transmission services."""
+
+        def add_torrent(service: ServiceCall) -> None:
+            """Add new torrent to download."""
+            if not (tm_client := _get_client(self.hass, service.data)):
+                raise ValueError("Transmission instance is not found")
+
+            torrent = service.data[ATTR_TORRENT]
+            if torrent.startswith(
+                ("http", "ftp:", "magnet:")
+            ) or self.hass.config.is_allowed_path(torrent):
+                tm_client.tm_api.add_torrent(torrent)
+                tm_client.api.update()
+            else:
+                _LOGGER.warning(
+                    "Could not add torrent: unsupported type or no permission"
+                )
+
+        def start_torrent(service: ServiceCall) -> None:
+            """Start torrent."""
+            if not (tm_client := _get_client(self.hass, service.data)):
+                raise ValueError("Transmission instance is not found")
+
+            torrent_id = service.data[CONF_ID]
+            tm_client.tm_api.start_torrent(torrent_id)
+            tm_client.api.update()
+
+        def stop_torrent(service: ServiceCall) -> None:
+            """Stop torrent."""
+            if not (tm_client := _get_client(self.hass, service.data)):
+                raise ValueError("Transmission instance is not found")
+
+            torrent_id = service.data[CONF_ID]
+            tm_client.tm_api.stop_torrent(torrent_id)
+            tm_client.api.update()
+
+        def remove_torrent(service: ServiceCall) -> None:
+            """Remove torrent."""
+            if not (tm_client := _get_client(self.hass, service.data)):
+                raise ValueError("Transmission instance is not found")
+
+            torrent_id = service.data[CONF_ID]
+            delete_data = service.data[ATTR_DELETE_DATA]
+            tm_client.tm_api.remove_torrent(torrent_id, delete_data=delete_data)
+            tm_client.api.update()
+
+        self.hass.services.async_register(
```

Comment:
We should leave registering the services in `__init__`

Reply:
I moved them to `__init__`. I also changed them to async now that we are requesting a coordinator update after the service call is executed.